### ISSUE TITLE
Set up a "nucleus similarity search" job type

### ISSUE THREAD

**Issue by waxlamp** [atlascope/atlascope#155]
The intent here is to select one of the nuclei and find other nuclei that are "similar" in some structural sense. The appropriate skimage function might be [structural_similarity](https://scikit-image.org/docs/stable/api/skimage.metrics.html#skimage.metrics.structural_similarity).

**Comment by mcovalt** [comment]
Here's the paper that introduces and describes the Structural Similarity Index Measure (SSIM):

Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). _Image quality assessment: From error visibility to structural similarity_. IEEE Transactions on Image Processing, 13, 600-612. https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf, [DOI:10.1109/TIP.2003.819861](https://doi.org/10.1109/TIP.2003.819861)

**Comment by mcovalt** [comment]
This turned out to be a very interesting topic. In brief, the SSIM compares three aspects of images: luminance, contrast, and "structure". Structure is what you get after you remove the luminance information from the images and normalize contrast. It is represented as an $(N-1)$​ dimensional unit vector for an image with $N$ pixels. The structural similarity between two images is the inner product of their two vectors.

Care must be taken to project the pixels to the same discrete space in cases such as ours where the two images are not identical. Windows of $M \times M$ pixels can be used to scan over the images. The $M \times M$ windows are associated with a discrete location and the windows are compared between images at their shared locations. Perhaps in our case the location could be $\langle x,y \rangle$ coordinates where origin is the center of the nucleus.

Vector similarity search is a really fascinating subject. We can view the problem as a nearest neighbor problem in high dimensions. PostGIS is actually the wrong tool for the job here. It doesn't support high dimensional vectors.

It is possible to represent the vectors in a [k-d tree](https://en.wikipedia.org/wiki/K-d_tree) using [PostgreSQL's SP-GiST](https://www.postgresql.org/docs/current/spgist-intro.html). The built-in PostgreSQL module to express high dimensional vectors is [cube](https://www.postgresql.org/docs/current/cube.html). This would be simplest since it requires no extra dependencies and should scale quite well. Since the vectors are unit vectors I _think_ it holds that the ordering provided by euclidean distance is equal to the ordering provided by their cosine similarity.

~I'd have to think about that last conjecture for a second to be sure though and that may make us a bit nervous.~ Alternatively, there's a purpose built tool for vector similarity searches: [pgvector](https://github.com/pgvector/pgvector). It offers inner product search for high dimensional vectors. There's even [an officially supported Django plugin](https://github.com/pgvector/pgvector-python#django). This has no constraint on using unit vectors so maybe we could do some interesting things like express window location as extra dimensions on the vector. However, real cosine similarity is much more expensive so this library uses approximate nearest neighbor search.

**tl;dr:** PostGIS is the wrong tool. Some properties of the problem allow us to use built-in SP-GiST/cube combo. Alternatively, a solution like pgvector can be used.

**Comment by mcovalt** [comment]
> Since the vectors are unit vectors I think it holds that the ordering provided by euclidean distance is equal to the ordering provided by their cosine similarity.

It seems this is true. [Here's a discussion behind the intuition](https://stats.stackexchange.com/questions/146221/is-cosine-similarity-identical-to-l2-normalized-euclidean-distance#comment280509_146279). With that in mind, it definitely makes sense to stick with built-in PostgreSQL SP-GiST based solutions.

**Comment by mcovalt** [comment]
With this analysis in mind, I'd like to clarify the problem and the proposed solution. My goal is really not to write so much, but there's a lot going on here. Sorry for the verbosity.

---

A scientist has many large images. The images are considered large because they contain so many pixels that the expected file size of each image is measured in gigabytes.

Some of these images capture the size and shape of many nuclei. A single image of this type will contain many nuclei. These images are analyzed to delineate the boundary of each individual nucleus (#160). A 2D polygon is provided per nucleus to express its spatial extent. The polygon is measured in unit "pixel".

**Problem:** The scientist would like to be able to select a single nucleus and get back a list nuclei across all analyzed images that are _structurally similar_ to the selected nucleus.

We define _structural similarity_ according to the [Structural Similarity Index Measure](https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf) (SSIM). The SSIM is a scalar between 0 and 1 that describes the correlation of two images. Note that the SSIM typically also measures luminance and contrast in addition to structure. Here, only the structure component of the measurement is considered.

The simplest form of the SSIM compares two images in their entirety. This requires assuming each image has the same number of pixels and structure is uniform across the entire image. Alternatively, the SSIM could be computed by averaging the _local_ SSIM across the images. The local SSIM between two images can be found by comparing the images via a _window_. For example, say a window of 7×7 pixels is used. The window crops out a 7×7 pixel section centered around each pixel of each image. Then, the 7×7 pixel crops are paired up, one from each image, where the center pixel of the 7×7 pixel crops represent the same location from each image. The SSIM is computed for each pair. This is the local SSIM. The average of all the local SSIM values is the overall SSIM comparing two images.

The structural similarity of two nuclei could be found using the windowed version of the SSIM as described above. Both nuclei would be cropped out of their original image. The cropped nuclei images would be aligned by placing their origin to the center of the crop and scaled/transformed so that the physical distance spanned by each pixel is identical between the images. The SSIM is then computed.

**Naive solution:** Provided a target nucleus, the above process is done to compare the target nucleus to every other nucleus in the database. The result list is sorted in descending order according to the computed SSIM. This is the list of all nuclei sorted by their structural similarity to the target nucleus.

But...

The SSIM computation is costly and the computation time of this job scales linearly with the number of nuclei in the database. The scientist claims this computation is infeasible to perform for each query.

Looking critically at the problem, we discover that the SSIM is actually the cosine of the angle between two unit vectors. The unit vector would be 48D for the 7×7 pixel windows.

**Proposed solution:** During nucleus delineation, a job would compute the 48D SSIM unit vector for each pixel of the aligned nuclei. The nuclei images would need to be normalized/scaled/transformed so that the pixel dimensions are constant across every nucleus image. The 48D vectors, their center pixel locations, and the vectors' associated nuclei is persisted in the database.  The above comments describe how PostgreSQL can be used to efficiently order all the nuclei according to their cosine similarity against a target nucleus using these vectors.
